node/core/prospective-parachains/src/lib.rs

```diff
@@ -0,0 +1,682 @@
+				// We are leaning hard on two assumptions here.
+				// 1. That the fragment tree never contains allowed relay-parents whose session for children
+				//    is different from that of the base block's.
+				// 2. That the max_pov_size is only configurable per session.
+				max_pov_size = Some(fragment_tree.scope().base_constraints().max_pov_size);
+			}
+		}
+	}
+
+	let _ = tx.send(match (head_data, relay_parent_info, max_pov_size) {
+		(Some(h), Some(i), Some(m)) => Some(PersistedValidationData {
+			parent_head: h,
+			relay_parent_number: i.number,
+			relay_parent_storage_root: i.storage_root,
+			max_pov_size: m as _,
+		}),
+		_ => None,
+	});
+}
+
+#[overseer::contextbounds(ProspectiveParachains, prefix = self::overseer)]
+async fn fetch_base_constraints<Context>(
+	ctx: &mut Context,
+	relay_parent: Hash,
+	para_id: ParaId,
+) -> JfyiErrorResult<Option<Constraints>> {
+	let (tx, rx) = oneshot::channel();
+	ctx.send_message(RuntimeApiMessage::Request(
+		relay_parent,
+		RuntimeApiRequest::StagingValidityConstraints(para_id, tx),
+	))
+	.await;
+
+	Ok(rx.await.map_err(JfyiError::RuntimeApiRequestCanceled)??.map(From::from))
+}
+
+#[overseer::contextbounds(ProspectiveParachains, prefix = self::overseer)]
+async fn fetch_upcoming_paras<Context>(
+	ctx: &mut Context,
+	relay_parent: Hash,
+) -> JfyiErrorResult<Vec<ParaId>> {
+	let (tx, rx) = oneshot::channel();
+
+	// This'll have to get more sophisticated with parathreads,
+	// but for now we can just use the `AvailabilityCores`.
+	ctx.send_message(RuntimeApiMessage::Request(
+		relay_parent,
+		RuntimeApiRequest::AvailabilityCores(tx),
+	))
+	.await;
+
+	let cores = rx.await.map_err(JfyiError::RuntimeApiRequestCanceled)??;
+	let mut upcoming = HashSet::new();
+	for core in cores {
+		match core {
+			CoreState::Occupied(occupied) => {
+				if let Some(next_up_on_available) = occupied.next_up_on_available {
+					upcoming.insert(next_up_on_available.para_id);
+				}
+				if let Some(next_up_on_time_out) = occupied.next_up_on_time_out {
+					upcoming.insert(next_up_on_time_out.para_id);
+				}
+			},
+			CoreState::Scheduled(scheduled) => {
+				upcoming.insert(scheduled.para_id);
+			},
+			CoreState::Free => {},
+		}
+	}
+
+	Ok(upcoming.into_iter().collect())
+}
+
+// Fetch ancestors in descending order, up to the amount requested.
+#[overseer::contextbounds(ProspectiveParachains, prefix = self::overseer)]
+async fn fetch_ancestry<Context>(
+	ctx: &mut Context,
+	relay_hash: Hash,
+	ancestors: usize,
+) -> JfyiErrorResult<Vec<RelayChainBlockInfo>> {
+	if ancestors == 0 {
+		return Ok(Vec::new())
+	}
+
+	let (tx, rx) = oneshot::channel();
+	ctx.send_message(ChainApiMessage::Ancestors {
+		hash: relay_hash,
+		k: ancestors,
+		response_channel: tx,
+	})
+	.await;
+
+	let hashes = rx.map_err(JfyiError::ChainApiRequestCanceled).await??;
+	let mut block_info = Vec::with_capacity(hashes.len());
+	for hash in hashes {
+		match fetch_block_info(ctx, relay_hash).await? {
```

Comment:
Should this be `hash` here?

```suggestion
		match fetch_block_info(ctx, hash).await? {
```